File: testdata/polyglot/fp/safe.js
// Safe-code plants for the FP measurement eval (see fp/safe.py header).
// PLANT-FP(id, CWE) marks the correct, non-vulnerable form of a weakness
// class — flagging it is a measured false positive.

const crypto = require('crypto');
const { execFile } = require('child_process');

function safeQuery(db, name) {
  // PLANT-FP(js-safe-sql, CWE-89): parameterized query, value is bound.
  return db.query('SELECT * FROM users WHERE name = $1', [name]);
}

function safeExec() {
  // PLANT-FP(js-safe-exec, CWE-78): execFile with a constant argument array,
  // no shell interpolation.
  return execFile('ls', ['-la'], () => {});
}

function safeHash(data) {
  // PLANT-FP(js-safe-hash, CWE-328): SHA-256 is a strong hash.
  return crypto.createHash('sha256').update(data).digest('hex');
}

module.exports = { safeQuery, safeExec, safeHash };

File: testdata/polyglot/fp/safe.py
# Safe-code plants for the FP measurement eval. Every construct here is the
# CORRECT, non-vulnerable form of a weakness class the scanners look for. A
# PLANT-FP(id, CWE) label marks code that must NOT be flagged for that CWE; if
# a profile flags it, that is a measured false positive (reported per-profile
# in docs/coverage.md), never a hidden one. Precision is measured, not asserted.

import subprocess
import sqlite3
import hashlib


def safe_sql(conn: sqlite3.Connection, username: str):
    # PLANT-FP(py-safe-sql, CWE-89): parameterized query — the value is bound,
    # not interpolated, so this is not SQL injection.
    cur = conn.cursor()
    cur.execute("SELECT * FROM users WHERE name = ?", (username,))
    return cur.fetchall()


def safe_shell():
    # PLANT-FP(py-safe-shell, CWE-78): a constant argument vector, no shell,
    # no user input — not command injection.
    return subprocess.check_output(["uptime"], shell=False)


def safe_hash(data: bytes):
    # PLANT-FP(py-safe-hash, CWE-328): SHA-256 for integrity is a strong hash,
    # not a weak-crypto finding.
    return hashlib.sha256(data).hexdigest()
